android/titanium/src/java/org/appcelerator/titanium/util/TiDownloadManager.java

```diff
@@ -50,19 +60,200 @@ public static TiDownloadManager getInstance()
 
 	protected TiDownloadManager()
 	{
-		handler = new Handler(this);
+		handler = new Handler(Looper.getMainLooper(), this);
 		threadPool = Executors.newFixedThreadPool(THREAD_POOL_SIZE);
 	}
 
+	/**
+	 * Downloads the content from URI asynchronously and provides the result via the given listener.
+	 * @param uri The URI to download the file/content from.
+	 * @param listener The listener to be invoked with the result of the download operation.
+	 */
 	public void download(URI uri, TiDownloadListener listener)
 	{
-		if (TiResponseCache.peek(uri)) {
+		if (TiResponseCache.peekFollowingRedirects(uri)) {
 			sendMessage(uri, MSG_FIRE_DOWNLOAD_FINISHED);
 		} else {
 			startDownload(uri, listener);
 		}
 	}
 
+	/**
+	 * Blocks the calling thread while download the file/content from the given URI.
+	 * <p>
+	 * Can be called on the main UI thread, but it is not recommended. The async download() method is preferred.
+	 * @param uri The URI to download the file/content from.
+	 * @return
+	 * Returns a stream to the file/content being downloaded.
+	 * <p>
+	 * Returns null if failed to download content or if given an invalid argument.
+	 */
+	public InputStream blockingDownload(URI uri)
+	{
+		// Validate.
+		if (uri == null) {
+			return null;
+		}
+
+		// First, check if the URI to the content/file has already been cached.
+		// Note: This is an optimization and avoids downloading the same content twice.
+		InputStream inputStream = null;
+		try {
+			URI cachedUri = TiResponseCache.fetchEndpointFollowingRedirects(uri);
+			if (cachedUri != null) {
+				inputStream = TiResponseCache.openCachedStream(cachedUri);
+				if (inputStream != null) {
+					return inputStream;
+				}
+			}
+		} catch (Exception ex) {
+		}
+
+		// If running on main UI thread, then block this thread while doing download on another thread.
+		// Note: Using "HttpUrlConnection" on UI thread will cause a "NetworkOnMainThreadException" to be thrown.
+		if (TiApplication.isUIThread()) {
+			// Anonymous class used to call this blockingDownload() method on another thread.
+			// Downloaded content will be made available via Titanium's "TiResponseCache".
+			class DownloadOperation implements Runnable
```

Comment:
Instead of creating an anonymous class, you should be able to do this:
```Java
public InputStream blockingDownload(final URI uri)
{
...
	// Perform the blocking download.
	try {
		Thread downloadThread = new Thread(new Runnable() {
			@Override
			public void run() {
				try (InputStream stream = blockingDownload(uri)) {
					if (stream != null) {
						KrollStreamHelper.pump(stream, null);
					}
				} catch (Exception ex) {
				}
			}
		});
		downloadThread.start();

		// wait for thread to finish
		downloadThread.join();
	} catch (Exception ex) {
	}
	...
...
```